### PR TITLE
Architecture lifecycles

### PR DESCRIPTION
**Pull request by ghost**

### CODE REVIEW

File: docker_images/architecture_manager/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,12 @@
+FROM dronesim/base
+
+ADD files/dependencies/*.jar /opt/felix/current/bundle/
+ADD files/bundles/*.jar /opt/felix/current/bundle/
+
+# You can override these at runtime, and you are encouraged to turn off debugger support in production
+ENV JVM_OPTIONS="-Xdebug -Xnoagent -Xrunjdwp:transport=dt_socket,address=8000,server=y,suspend=n -Dgosh.args=--nointeractive"
+
+CMD exec java $JVM_OPTIONS -jar bin/felix.jar 
+#CMD ["java", "-Dgosh.args=--nointeractive", "-jar","bin/felix.jar"]
```
Comment (jjkester) [inline docker_images/architecture_manager/Dockerfile:10]: Left over from testing?

File: implementation/architecture-event-controller/pom.xml [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+        <artifactId>dronessimulator</artifactId>
+        <groupId>org.inaetics</groupId>
+        <version>0.1</version>
+    </parent>
+    <modelVersion>4.0.0</modelVersion>
+
+    <packaging>bundle</packaging>
+
+    <groupId>org.inaetics.dronessimulator</groupId>
+    <artifactId>architecture-event-controller</artifactId>
```
Comment (jjkester) [inline implementation/architecture-event-controller/pom.xml:17]: Maybe add a `<name>`?

File: implementation/architecture-event-controller/pom.xml [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+                    <instructions>
+                        <Export-Package>org.inaetics.dronessimulator.architectureevents.*</Export-Package>
+                        <Bundle-Activator>org.inaetics.dronessimulator.architectureevents.Activator</Bundle-Activator>
+                        <Embed-Transitive>true</Embed-Transitive>
+                    </instructions>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>
```
Comment (jjkester) [inline implementation/architecture-event-controller/pom.xml]: Newline

File: implementation/common/src/main/java/org/inaetics/dronessimulator/common/protocol/ArchitectureMessage.java [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+
+import org.inaetics.dronessimulator.common.architecture.SimulationAction;
+
+import java.util.Collections;
+import java.util.List;
+
+public class ArchitectureMessage extends ProtocolMessage {
+    private SimulationAction action = null;
+
+
```
Comment (jjkester) [inline implementation/common/src/main/java/org/inaetics/dronessimulator/common/protocol/ArchitectureMessage.java]: Superfluous line

File: implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java [hunk 5]
```diff
@@ -28,6 +27,11 @@
     void unregister(Instance instance) throws IOException;
 
     /**
+     * Updates the properties associated with an instance. The instance must exist!
+     */
+     Instance updateProperties(Instance instance, Map<String, String> properties) throws IOException;
```
Comment (jjkester) [inline implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java]: At first glance this interface seems a bit weird. I personally would try to make the use case clear in the method documentation.

File: implementation/drone/components/engine/src/java/org/inaetics/dronessimulator/drone/components/engine/Engine.java [hunk 6]
```diff
@@ -1,18 +1,18 @@
...
 import org.inaetics.dronessimulator.drone.droneinit.DroneInit;
 import org.inaetics.dronessimulator.pubsub.api.publisher.Publisher;
-import org.inaetics.dronessimulator.pubsub.api.subscriber.Subscriber;
 
 import java.io.IOException;
 
-/**
- * Created by mart on 17-5-17.
- */
+
```
Comment (jjkester) [inline implementation/drone/components/engine/src/java/org/inaetics/dronessimulator/drone/components/engine/Engine.java]: Superfluous newline

File: implementation/drone/components/engine/src/java/org/inaetics/dronessimulator/drone/components/engine/Engine.java [hunk 7]
```diff
@@ -67,7 +67,8 @@ public void changeAcceleration(D3Vector input_acceleration){
         try {
             m_publisher.send(MessageTopic.MOVEMENTS, msg);
         } catch (IOException e) {
-            System.out.println("Exception");
+            logger.fatal(e);
+            e.printStackTrace();
```
Comment (jjkester) [inline implementation/drone/components/engine/src/java/org/inaetics/dronessimulator/drone/components/engine/Engine.java]: Is it really that fatal that we need to print a separate stack trace?

File: implementation/drone/components/gps/src/java/org/inaetics/dronessimulator/drone/components/gps/GPS.java [hunk 8]
```diff
@@ -28,7 +31,8 @@ public void start() {
         try {
             this.m_subscriber.addTopic(MessageTopic.STATEUPDATES);
         } catch (IOException e) {
-            System.out.println("IO Exception add Topic"); // todo logging
+            logger.fatal(e);
+            e.printStackTrace();
```
Comment (jjkester) [inline implementation/drone/components/gps/src/java/org/inaetics/dronessimulator/drone/components/gps/GPS.java]: See previous stacktrace comment

File: implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Activator.java [hunk 9]
```diff
@@ -15,14 +16,19 @@ public void init(BundleContext bundleContext, DependencyManager dependencyManage
...
                 .setImplementation(Radar.class)
+                .setCallbacks("init", "start", "stop", "destroy")
                 .add(createServiceDependency()
                         .setService(DroneInit.class)
                         .setRequired(true)
                 )
                 .add(createServiceDependency()
                         .setService(Subscriber.class)
                         .setRequired(true)
-                ).setCallbacks("init", "start", "stop", "destroy")
```
Comment (jjkester) [inline implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Activator.java]: Sure that the callbacks do not need to be set? I would think they are there for a reason.

File: implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Activator.java [hunk 10]
```diff
@@ -15,14 +16,19 @@ public void init(BundleContext bundleContext, DependencyManager dependencyManage
...
+                )
+                .add(createServiceDependency()
+                    .setService(ArchitectureEventController.class)
+                    .setRequired(true)
+                )
         );
     }
 }
```
Comment (jjkester) [inline implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Activator.java:31]: Newline

File: implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Radar.java [hunk 11]
```diff
@@ -30,10 +38,17 @@ public void start() {
         try {
             this.m_subscriber.addTopic(MessageTopic.STATEUPDATES);
         } catch (IOException e) {
-            System.out.println("IO Exception add Topic");
+            logger.fatal(e);
+            e.printStackTrace();
```
Comment (jjkester) [inline implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Radar.java]: See previous stacktrace comment

File: implementation/gameengine/gamestate-manager/src/main/java/org/inaetics/dronessimulator/gameengine/gamestatemanager/GameStateManager.java [hunk 12]
```diff
@@ -68,6 +73,10 @@ public GameEntity getById(Integer id) {
      * Starts the game state manager service.
```
Comment (jjkester) [inline implementation/gameengine/gamestate-manager/src/main/java/org/inaetics/dronessimulator/gameengine/gamestatemanager/GameStateManager.java:73]: Document arch controller action

**Review by jjkester — CHANGES_REQUESTED** [review]
In general a *lot* of documentation is still needed. It would be advisable to document at least everything in classes/packages that I already checked/improved so there is no need to go over all the code again.

File: implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java (hunk 5 above)
Comment (ghost) [reply implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java]: Is needed for architecture management. The current lifecycle state is published with discovery to make use of the persistent storage. However, this state changes and so the properties have to be updated

File: implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Activator.java (hunk 9 above)
Comment (ghost) [reply implementation/drone/components/radar/src/java/org/inaetics/dronessimulator/drone/components/radar/Activator.java]: They do not have to be set, these are the default ones.

**Comment by ghost** [comment]
Changes will be added to sonarqube-cleanup branch. Already fixed a lot of issues in that specific branch. Would you will be willing to accept this branch so the commits will not have to be checked again?

Documentation will also be added to this branch.

**Comment by ghost** [comment]
Requested changes are at  	9ab9f13

File: implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java (hunk 5 above)
Comment (jjkester) [reply implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java]: (...) _in the method documentation_

**Comment by jjkester** [comment]
Better to just reference this PR from the sonarqube-cleanup PR instead of just accepting this one.

**Comment by ghost** [comment]
Changed the base from sonarqube-cleanup PR to this branch. This will allow the reviewer to just see the new changes/diff instead of having to go over all other reviewed code from this branch as well. When that PR is accepted, the requested changes will be in this PR so we can accept again.

**Review by Langstra — APPROVED** [review]

File: implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java (hunk 5 above)
Comment (jjkester) [reply implementation/discovery/api/src/main/java/org/inaetics/dronessimulator/discovery/api/Discoverer.java]: #89

**Comment by ghost** [comment]
Gefixed! Mag hij nu naar master? :P

**Review by jjkester — APPROVED** [review]

**Pull request merged**